src/library/scala/collection/mutable/HashTable.scala

```
@@ -147,6 +151,24 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
       resize(2 * table.length)
   }
 
+  /** Find entry with given key in table, or add new one if not found.
+   *  May be somewhat faster then `findEntry`/`addEntry` pair as it
+   *  computes entry's hash index only once.
+   *  Returns entry found in table or null.
+   *  Method `createNewEntry` must be implemented in any subclass/subtrait who calls this.
+   */
+  protected def findOrAddEntry[B](key: A, value: B): Entry = {
+    val h = index(elemHashCode(key))
+    val e = findEntry0(key, h)
+    if (e ne null) e else { addEntry0(createNewEntry(key, value), h); null }
+  }
+
+  /** Creates new entry to be immediately inserted into the hashtable.
+   *  This method must be implemented for any class who calls `findOrAddEntry`.
+   *  Ideally, it should be abstract, but this will break source compatibility.
+   */
+  protected def createNewEntry[B](key: A, value: B): Entry = ???
```

I'm not sure if having a default implementation which silently passes compilation to retain source compatibility and then breaks at runtime is the best thing to do.
Granted, since only `HashMap`s and `LinkedHashSet`s currently use the `findOrAddEntry` which calls this method, and these classes implement `createNewEntry`, it should not break the existing code.

However, since I can't think of any other solution, I recommend we keep this.